### PR TITLE
Dash refactor

### PR DESCRIPTION
**Pull request by serge-gaia**
[#1914: Save dashboard tables settings saved session ](https://kanboard.dpaw.wa.gov.au/?controller=task&action=show&task_id=1914&project_id=24).

### CODE REVIEW

File: wildlifelicensing/apps/customer_management/views/customer.py [hunk 1]
```diff
@@ -144,6 +147,81 @@ def _build_data(self):
...
+    def returns_filters(self):
+        # no filters
+        return {}
+
+    @property
+    def get_returns_session_data(self):
+        # no session
+        return {}
+
     def get(self, request, *args, **kwargs):
```
Comment (serge-gaia) [inline wildlifelicensing/apps/customer_management/views/customer.py]: This is the base class that all Tables<...> will extend and override all the `@properties`

File: wildlifelicensing/apps/dashboard/static/wl/js/dash_tables.js [hunk 2]
```diff
@@ -3,11 +3,12 @@ define(
         'jQuery',
```
Comment (serge-gaia) [inline wildlifelicensing/apps/dashboard/static/wl/js/dash_tables.js:3]: Changed this script to apply filters.

File: wildlifelicensing/apps/dashboard/views/base.py [hunk 3]
```diff
@@ -150,62 +150,304 @@ def get_context_data(self, **kwargs):
         return super(DashboardTreeViewBase, self).get_context_data(**kwargs)
 
 
-class TableBaseView(TemplateView):
+class TablesBaseView(TemplateView):
+    """
```
Comment (serge-gaia) [inline wildlifelicensing/apps/dashboard/views/base.py:154]: This is the base class that all Tables<...> will extend

File: wildlifelicensing/apps/dashboard/views/base.py [hunk 4]
```diff
@@ -150,62 +150,304 @@ def get_context_data(self, **kwargs):
...
+            self.set_filter(result, name, values)
+
+        # apply session if no query parameters
+        request = self.request if hasattr(self, 'request') else None
+        if request and not request.GET:
+            session_data = self.get_returns_session_data
+            if session_data:
+                self.update_with_session_data(result, session_data)
+        return result
+
```
Comment (serge-gaia) [inline wildlifelicensing/apps/dashboard/views/base.py:342]: subclasses override the @properties

File: wildlifelicensing/apps/dashboard/views/base.py [hunk 5]
```diff
@@ -252,6 +494,37 @@ class DataTableBaseView(LoginRequiredMixin, BaseDatatableView):
         }
     }
 
+    SESSION_SAVE_SETTINGS = True  # Set to True if you want to save order/search in the session
```
Comment (serge-gaia) [inline wildlifelicensing/apps/dashboard/views/base.py:497]: This is were all the session things happen

File: wildlifelicensing/apps/dashboard/views/base.py [hunk 6]
```diff
@@ -150,62 +150,304 @@ def get_context_data(self, **kwargs):
...
-                    'licenceType': {
-                        'values': licence_types,
-                    },
-                },
-                'ajax': {
-                    'url': ''
-                }
+        # dataTable options
+        dt_options = {}
+        if session_data.get('pageLength'):
```
Comment (tony-gaia) [inline wildlifelicensing/apps/dashboard/views/base.py:416]: should this be `if session_data.get('pageLength') is not None:`?

**Pull request merged**